### DISCUSSION TITLE
Pointset Boolean Difference

### DISCUSSION THREAD

**Discussion by petrasvestartas** [CGAL/cgal#6405] [Q&A] [answered]
Hi,

I would like to ask which method I should use to remove Closest Points from another pointset?
Should I use octree search or a similar method? 
Both pointset have very large number of points.

This is an illustration what I am searching for:
![Screenshot_20220311-211233_Samsung Notes](https://user-images.githubusercontent.com/18013985/157956009-30e8944f-828b-4707-a062-92b44eac71b1.jpg)

**Comment by afabri** [comment]
2D or 3D?

**Reply by petrasvestartas** [reply]
Dear @afabri ,

In 3d, we are scanning timber roof shingles with stationary LIDAR in epfl. The goal is to crop one set from the other set when a new object is placed in the scanning scene.

**Reply by petrasvestartas** [reply]
Dear @afabri ,

Should I use Octree for this problem? 
https://doc.cgal.org/latest/Orthtree/index.html

**Comment by afabri** [comment]
I guess your other question is for the same problem.  Do you want to program something like "If in a small sphere around a yellow point are some blue points, the yellow point should be removed"?
Can you share some representative point sets?

**Reply by petrasvestartas** [reply]
Dear @afabri ,


Yes the other issue is the same. 
This is correct: "If in a small sphere around a yellow point are some blue points, the yellow point should be removed"


I will try to use your example posted in the other issue.


For CGAL point-set visualization do I need to install additional libraries like QT or this example would work if I just download the latest cgal release?
https://doc.cgal.org/latest/Point_set_3/Point_set_3_2draw_point_set_3_8cpp-example.html#a7

I also attach the point-sets.

[PointSets.zip](https://github.com/CGAL/cgal/files/8443236/PointSets.zip)

**Comment by afabri** [comment, marked as answer]
For the drawing you need Qt.

**Comment by afabri** [comment]
I do not really understand what you want to achieve. When I visualize the two point sets drawn in red and blue there is a zone A where red is above blue, a zone B where blue is above red, and a zone C with only blue points.
![image](https://user-images.githubusercontent.com/3263539/162246976-bdeb6666-ac71-4fe0-a687-2fa2a6734a95.png)

**Reply by petrasvestartas** [reply]
Dear @afabri ,

I finally managed to run this example.

Red - 1st Scan
Yellow - 2nd Scan
Black - Closest Point Distance between the two.

![image](https://user-images.githubusercontent.com/18013985/164088334-76b37064-776c-485d-816a-bc3fcb07c39e.png)


Not sure if the code is completely right, but I get the result I was searching for.
It shows that we need to change our experiment a little bit. But it is amazing how fast is CGAL to compute the point search...

This is the code I used:
```cpp
#pragma once
#include "cgal.h"

#include <CGAL/Simple_cartesian.h>
#include <CGAL/Kd_tree.h>
#include <CGAL/point_generators_3.h>
#include <CGAL/algorithm.h>
#include <CGAL/Fuzzy_sphere.h>
#include <CGAL/Search_traits_3.h>
typedef CGAL::Simple_cartesian<double>                K;
typedef K::Point_3                                    Point_d;
typedef CGAL::Search_traits_3<K>                      Traits;
typedef CGAL::Fuzzy_sphere<Traits>                    Fuzzy_circle;
typedef CGAL::Kd_tree<Traits>                         Tree;

inline void cgal_pointcloud_boolean_difference(CGAL::Point_set_3<IK::Point_3>& cgal_cloud_0, CGAL::Point_set_3<IK::Point_3>& cgal_cloud_1, CGAL::Point_set_3<IK::Point_3>& cgal_cloud_result, double radius = 2)
{

    Tree tree;

    for (auto& p : cgal_cloud_0.points()) {
        Point_d point(p.x(), p.y(), p.z());
        tree.insert(point);
    }
	

    cgal_cloud_result.reserve(cgal_cloud_1.size());
	
    for (auto& point : cgal_cloud_1.points()) {

        Point_d center(point.hx(), point.hy(), point.hz());

        Fuzzy_circle default_range(center, radius);
        std::list<Point_d> result;
        tree.search(std::back_inserter(result), default_range);

        if (result.size() == 0) {
            cgal_cloud_result.insert(point);
         }
        
    }
 
} 
```

**Comment by afabri** [comment]
Just three remarks.  No need for create `point` from `p` and later `center` from `point`.   They are of the same type.  Also instead of filling `result` with potentially many points you better write a container that can hold the first point and then throws an exception. Finally, but that is minor, better use `std::list::empty()` than checking if the size equals zero.

**Reply by petrasvestartas** [reply]
Thank you very much.

I hesitated about these points whether they are correct or not.
Would it be possible to show how to do in code the following ?

`Also instead of filling result with potentially many points you better write a container that can hold the first point and then throws an exception. `

I am trying to create a new pointset with points that are more further than the given search radius.
